### DISCUSSION TITLE
How to upgrade a Python minor version, eg `3.8.x` -> `3.9.x`?

### DISCUSSION THREAD

**Discussion by fyrak1s** [gobolinux/Documentation#31] [Q&A]
### What is the proper way to update a Python minor version, like `3.8.x` to `3.9.x`?

My steps are like this:
```sh
NewVersion Python 3.9.20
[...]
Compile Python 3.9.20
```

It compiles and installs fine. With this exception:
```
SymlinkProgram: Conflict: /System/Aliens/PIP/3.8/bin/pip3
```

`python3 --version` correctly shows `Python 3.9.20`.

But upon trying to install a meson recipe, eg `Mesa`, I am getting:
```
Compile: This is a meson type recipe and you don't seem to have meson installed.
```

This seems to be due to Alien (since meson is installed via PIP3/Aliens). So I am trying to reinstall meson:
```
Alien --install PIP3:meson
```
This works actually:
```
which meson
/System/Aliens/PIP/3.9/bin/meson
```
However I still cannot compile meson recipes.
Compile would still not find meson!

Also upon re-running `Alien --install PIP3:meson` I am getting:
```
Requirement already satisfied: meson in /System/Aliens/PIP/3.8/lib/python3.8/site-packages (1.5.2)
```
which is referencing python 3.8 again (?).

To make it more complicated under the `/System/Aliens/PIP/3.9/lib/` subfolder (**note: 3.9!!**) we are getting `python3.8/site-packages` again. Full path:
```
/System/Aliens/PIP/3.9/lib/python3.8/site-packages
```

It seems Aliens/PIP3(?) has to be updated to work with the new python minor?

However, if I understand the python recipe and its [Alien compatibility patch](https://github.com/gobolinux/Recipes/blob/master/Python/3.8.1/01-search-SLL.patch.in) correctly, Aliens should actually re-use the packages installed under Python3.8 with PIP3:

> We let a program to import modules from older Python versions that share the same major number.

This does not seem to work though?!
(I did update `PYTHON_MINOR=7` to `PYTHON_MINOR=9` in the patch btw)

---

Further, I wanted to make meson recipes Compile again, so I did a
```
SymlinkProgram Python 3.8.1
```
This did not fix the problem, still getting:
"This is a meson type recipe and you don't seem to have meson installed."

The only way to "fix' it was to 
```
RemoveProgram Python 3.9.20
```
in the terminal.

I was not brave enough to do it the other way around and delete Python 3.8.1 to see if that would make it work 😅

### I'd be grateful for any help or pointers!

PS: Python >= 3.9 is actually required to successfully build Mesa nowadays. So that is why I am trying to update python in the first place.

**Comment by rubyFeedback** [comment]
I can not answer the question as such; compiling new python versions will typically install
e. g. "pip3" under bin/ and so forth. I guess pip is the main way to install new python
packages; the LFS/BLFS projects all show installations via pip + wheel nowadays,
such as for packaging:

https://www.linuxfromscratch.org/blfs/view/svn/general/python-modules.html#packaging

    pip3 wheel -w dist --no-build-isolation --no-deps --no-cache-dir $PWD
    pip3 install --no-index --find-links dist --no-user packaging

I do not know how Alien works under the hood, but I also use a versioned AppDir
on LFS-centric computer systems here, using the traditional /usr/ prefix as main
setup, so for instance, Python would then reside at:

    /home/Programs/Python/3.13.2/

(I use /home/Programs/ rather than /Programs/ - while the latter is much more
elegant and logical, I started to use /home/Programs/ largely because I may
want to backup it more easily, while also having used systems such as fedora,
which tend to use different partitions, so I found it easier to put all I needed
into /home/ actually. /Programs/ is much more logical though. I use the
shell environment variable $MY_PROGRAMS to indicate which programs 
directory I use here.)

After I compiled this python version and changed the symlink at /usr/bin/python,
I can install the other python addons just fine, including meson. I then end up
with a meson binary e. g. here:

    /home/Programs/Python/3.13.2/bin/meson
    /home/Programs/Python/3.11.1/bin/meson # if I use python 3.11.x

And these work fine the moment I installed them, so whatever Alien is doing,
I think it is doing something wrong since the above approach works for me
as-is on non-GoboLinux systems. I don't know the internals though, so I
could be wrong.

From the error message I would assume that the python binary is confused
about which path to use. Which is a bit odd though - how does python 
check which path to use? With Ruby the above works even simpler, as I
don't even need e. g. /usr/lib/ruby/ or something like that - everything
resides in the versioned AppDir as-such.

> I was not brave enough to do it the other way around and delete Python 3.8.1 to see if that would make it work 

I think you can just temporarily rename it rather than delete it. I use that
approach with my ruby scripts; I have a variant that removes broken 
symlinks, so I just rename the directory, remove broken symlinks, then
rename the directory back to the original name again.

By the way, is Alien actually really needed? I understand some of the rationale,
but I kind of ended up using mostly just one version. In ruby I tend to use
the latest git checkout; in python I either use 3.13.x, or for some legacy
reasons also have 3.11.x available. I can always re-install everything
easily via "pip install xyz" such as "pip install jinja2" or so.

PS: I also noticed mesa is a bit outdated in GoboLinux:

https://github.com/gobolinux/Recipes/tree/master/Mesa

https://www.linuxfromscratch.org/blfs/view/svn/x/mesa.html

Mesa kind of got more dependencies now - llvm, spirv and so 
forth. spirv is actually easy if llvm is available. Thankfully
LLVM in GoboLinux is not too terribly old - I guess you 
updated it already:

https://github.com/gobolinux/Recipes/tree/master/LLVM/19.1.0

Could try 20.1.1 - from my limited testing on a non-gobolinux
system, there should not be too much difference here between
19.x and 20.x

https://www.linuxfromscratch.org/blfs/view/svn/general/llvm.html
